src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingVisitor.cs

```diff
@@ -212,10 +212,56 @@ public override void VisitMarkupTagHelperElement(MarkupTagHelperElementSyntax no
 
             Visit(node.EndTag);
 
-            static bool IsComponentTagHelperNode(MarkupTagHelperElementSyntax node)
+            static bool CausesGeneratedCSharpIndentation(MarkupTagHelperElementSyntax node)
             {
-                return node.TagHelperInfo?.BindingResult?.Descriptors?.Any(
-                    d => d.IsComponentOrChildContentTagHelper()) ?? false;
+                var tagHelperInfo = node.TagHelperInfo;
+
+                if (tagHelperInfo is null)
+                {
+                    return false;
+                }
+
+                var descriptors = tagHelperInfo.BindingResult?.Descriptors;
+                if (descriptors is null)
+                {
+                    return false;
+                }
+
+                // If there are any child tag helpers that match properties of this tag helper, that are RenderFragments,
+                // then it means this specific node won't actually cause a change in indentation.
+                // For example, the following two bits of Razor generate identical C# code, even though the code block is
+                // nested in a different number of tag helper elements:
+                //
+                // <Component>
+                //     @if (true)
+                //     {
+                //     }
+                // </Component>
+                //
+                // and
+                //
+                // <Component>
+                //     <ChildContent>
+                //         @if (true)
+                //         {
+                //         }
+                //     </ChildContent>
+                // </Component>
+                //
+                // This code will not count "Component" as causing indentation because it has a child component called "ChildContent"
+                // and a property called "ChildContent".
+
+                // Get the child components, if any
+                var childComponents = node.ChildNodes().OfType<MarkupTagHelperElementSyntax>()
+                    .Where(n => n.TagHelperInfo?.BindingResult.Descriptors.Any(d => d.IsComponentOrChildContentTagHelper()) ?? false)
```

Comment:
@SteveSandersonMS is there something ont he binding result we can look into to seeif a child component was fulfilled? It hurts that @davidwengier has to reapply this matching logic here 😄

Reply:
I'm gonna re-do this to look at the parent, not the children, and I think it will end up a bit nicer.